### PR TITLE
Add event logging to libpod, even display to podman

### PR DESCRIPTION
**Pull request by baude**
In lipod, we now log major events that occurr.  These events
can be displayed using the `podman events` command. Each
event contains:

* Type (container, image, volume, pod...)
* Status (create, rm, stop, kill, ....)
* Timestamp in RFC3339Nano format
* Name (if applicable)
* Image (if applicable)

The format of the event and the varlink endpoint are to not
be considered stable until cockpit has done its enablement.

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

File: docs/podman-events.1.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+## SYNOPSIS
+**podman events** [*options*]
+
+## DESCRIPTION
+
+Monitor and print events that occur in Podman. Each event will include a timestamp,
+a type, a status, name (if applicable), and image (if applicable).
+
+Podman events will report the following event types:
```
Comment (rhatdan) [inline docs/podman-events.1.md]: I think the man page should be rewritten for reporting events by category.

This looks like I could get a checkpoint event for a volume.

File: docs/podman-events.1.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+The *since* and *until* values can be RFC3339Nano time stamps or a Go duration string such as 10m, 5h. If no
+*since* or *until* values are provided, only new events will be shown.
+
+## EXAMPLES
+
+Showing podman events
+```
+$ podman events
+2019-03-02 10:33:42.235316587 -0600 CST image inspect caf27325b298 docker.io/library/alpine:latest
```
Comment (rhatdan) [inline docs/podman-events.1.md]: Does docker log inspect events?

File: libpod/events/events.go [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+// Recycle checks if the event log has reach a limit and if so
+// renames the current log and starts a new one.  The remove bool
+// indicates the old log file should be deleted.
+func (e *Event) Recycle(path string, remove bool) error {
+	return errors.New("not implemented")
+}
+
+// ToJSONString returns the event as a json'ified string
+func (e *Event) ToJSONString() string {
+	b, _ := json.Marshal(e)
```
Comment (rhatdan) [inline libpod/events/events.go]: Should at least log errors.

File: libpod/events/events.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+	return fmt.Sprintf("%s\n", b)
+}
+
+// ToHumanReadable returns human readable event as a formatted string
+func (e *Event) ToHumanReadable() string {
+	var humanFormat string
+
+	switch e.Type {
+	case Container, Volume, Pod:
```
Comment (rhatdan) [inline libpod/events/events.go]: Volumes do not have images associated with them.

File: libpod/events/events.go [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+		humanFormat = fmt.Sprintf("%s %s %s %s (image=%s, name=%s)", e.Time, e.Type, e.Status, e.ID, e.Image, e.Name)
+	case Image:
+		humanFormat = fmt.Sprintf("%s %s %s %s %s", e.Time, e.Type, e.Status, e.ID, e.Name)
+	}
```
Comment (rhatdan) [inline libpod/events/events.go]: Should log an assert if you get an unsatisfied e.Type.

File: libpod/events/events.go [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+	if err := json.Unmarshal([]byte(event), &e); err != nil {
+		return nil, err
+	}
+	return &e, nil
+
+}
+
+// ToString converts a Type to a string
+func (t Type) ToString() string {
+	return fmt.Sprintf("%s", t)
```
Comment (rhatdan) [inline libpod/events/events.go]: Can you return t?  Or string(t)

File: libpod/events/events.go [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+// ToString converts a status to a string
+func (s Status) ToString() string {
+	return fmt.Sprintf("%s", s)
```
Comment (rhatdan) [inline libpod/events/events.go]: Can you return s or string(s)

File: libpod/pod_api.go [hunk 8]
```diff
@@ -447,5 +449,6 @@ func (p *Pod) Inspect() (*PodInspect, error) {
 		},
 		Containers: podContainers,
 	}
+	defer p.newPodEvent(events.Inspect)
```
Comment (rhatdan) [inline libpod/pod_api.go]: Why do you call these as defer rather then just calling them?

File: libpod/runtime.go [hunk 9]
```diff
@@ -1077,3 +1094,8 @@ func (r *Runtime) ImageRuntime() *image.Runtime {
 func (r *Runtime) SystemContext() *types.SystemContext {
 	return r.imageContext
 }
+
+// GetEventLogFilePath returns the path for the event log file in string form
+func (r *Runtime) GetEventLogFilePath() string {
```
Comment (rhatdan) [inline libpod/runtime.go]: Remove Get, just call it EventLogFilePath()

**Comment by rhatdan** [comment]
Missing completions.

File: libpod/container_api.go [hunk 10]
```diff
@@ -125,7 +127,7 @@ func (c *Container) StartAndAttach(ctx context.Context, streams *AttachStreams,
 		}
 		close(attachChan)
 	}()
-
+	defer c.newContainerEvent(events.StartAndAttach)
```
Comment (mheon) [inline libpod/container_api.go]: This should be two separate events - a start and an attach

File: libpod/container_api.go [hunk 11]
```diff
@@ -351,7 +353,7 @@ func (c *Container) Exec(tty, privileged bool, env, cmd []string, user, workDir
 	if err := c.save(); err != nil {
 		logrus.Errorf("Error removing exec session %s from container %s state: %v", sessionID, c.ID(), err)
 	}
-
+	c.newContainerEvent(events.Exec)
```
Comment (mheon) [inline libpod/container_api.go]: This needs to go a lot further up, right after we start the exec but before we unlock and wait for it. Still needs to not be a defer.

File: libpod/container_api.go [hunk 12]
```diff
@@ -542,12 +545,13 @@ func (c *Container) Inspect(size bool) (*inspect.ContainerInspectData, error) {
 	if err != nil {
 		return nil, errors.Wrapf(err, "error getting graph driver info %q", c.ID())
 	}
-
+	defer c.newContainerEvent(events.Inspect)
```
Comment (mheon) [inline libpod/container_api.go]: Does this count as an event? We're not modifying the container at all

File: libpod/container_api.go [hunk 13]
```diff
@@ -542,12 +545,13 @@ func (c *Container) Inspect(size bool) (*inspect.ContainerInspectData, error) {
...
 	return c.getContainerInspectData(size, driverData)
 }
 
 // Wait blocks until the container exits and returns its exit code.
 func (c *Container) Wait() (int32, error) {
+	defer c.newContainerEvent(events.Wait)
```
Comment (mheon) [inline libpod/container_api.go]: I think you're doubling up here - WaitWithInterval already adds a Wait event

File: libpod/container_api.go [hunk 14]
```diff
@@ -772,7 +777,7 @@ func (c *Container) Refresh(ctx context.Context) error {
 			return err
 		}
 	}
-
+	defer c.newContainerEvent(events.Refresh)
```
Comment (mheon) [inline libpod/container_api.go]: Drop this. Refresh is deprecated, we shouldn't be updating it.

File: libpod/container_api.go [hunk 15]
```diff
@@ -815,6 +820,6 @@ func (c *Container) Restore(ctx context.Context, options ContainerCheckpointOpti
 			return err
 		}
 	}
-
+	defer c.newContainerEvent(events.CheckpointRestore)
```
Comment (mheon) [inline libpod/container_api.go]: Maybe rename from CheckpointRestore to RestoreFromCheckpoint (or just Restore)?

File: libpod/container_internal.go [hunk 16]
```diff
@@ -1022,7 +1023,7 @@ func (c *Container) restartWithTimeout(ctx context.Context, timeout uint) (err e
 			return err
 		}
 	}
-
+	defer c.newContainerEvent(events.Restart)
```
Comment (mheon) [inline libpod/container_internal.go]: Do we want a dedicated restart event? I'm fine having restart generate separate Stop and Start events

File: libpod/events.go [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+	e.Name = v.Name()
+	e.Type = events.Volume
+	if err := e.Write(v.runtime.config.EventsLogFilePath); err != nil {
+		logrus.Infof("unable to write event log to %s", v.runtime.config.EventsLogFilePath)
+	}
+}
+
+// GetEventsTail is a wrapper function for everyone to begin tailing the events log
+// with options
+func (r *Runtime) GetEventsTail(fromStart, stream bool) (*tail.Tail, error) {
```
Comment (mheon) [inline libpod/events.go]: What is a tail.Tail?

I would really prefer we do a chan(events.Event)

File: libpod/events.go [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+	e.Type = events.Pod
+	if err := e.Write(p.runtime.config.EventsLogFilePath); err != nil {
+		logrus.Infof("unable to write event log to %s", p.runtime.config.EventsLogFilePath)
+	}
+}
+
+// newVolumeEvent creates a new event for a libpod volume
+func (v *Volume) newVolumeEvent(status events.Status) {
+	e := events.NewEvent(status)
+	e.ID = v.Name()
```
Comment (mheon) [inline libpod/events.go]: Leave this blank - volumes don't have an ID and we can't pretend they do

File: libpod/events.go [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+	e.Name = v.Name()
+	e.Type = events.Volume
+	if err := e.Write(v.runtime.config.EventsLogFilePath); err != nil {
+		logrus.Infof("unable to write event log to %s", v.runtime.config.EventsLogFilePath)
```
Comment (mheon) [inline libpod/events.go]: Errors should be errorf

File: libpod/events.go [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// newPodEvent creates a new event for a libpod pod
+func (p *Pod) newPodEvent(status events.Status) {
+	e := events.NewEvent(status)
+	e.ID = p.ID()
+	e.Name = p.Name()
+	e.Type = events.Pod
+	if err := e.Write(p.runtime.config.EventsLogFilePath); err != nil {
+		logrus.Infof("unable to write event log to %s", p.runtime.config.EventsLogFilePath)
```
Comment (mheon) [inline libpod/events.go]: Errorf

File: libpod/events.go [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+
+// newContainerEvent creates a new event based on a container
+func (c *Container) newContainerEvent(status events.Status) {
+	e := events.NewEvent(status)
+	e.ID = c.ID()
+	e.Name = c.Name()
+	e.Image = c.config.RootfsImageName
+	e.Type = events.Container
+	if err := e.Write(c.runtime.config.EventsLogFilePath); err != nil {
+		logrus.Infof("unable to write event log to %s", c.runtime.config.EventsLogFilePath)
```
Comment (mheon) [inline libpod/events.go]: Errorf

File: libpod/events/events.go (hunk 3 above)
Comment (mheon) [reply libpod/events/events.go]: This should return an error - we don't want to write when we failed to produce a sane error

Comment (mheon) [reply libpod/events/events.go]: Sane event, even

File: libpod/events/events.go (hunk 6 above)
Comment (mheon) [reply libpod/events/events.go]: string(t) ought to be fine.

Also rename to `String()` - it's a generally-used interface in Golang that a lot of methods accept

File: libpod/events/events.go [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+		humanFormat = fmt.Sprintf("%s %s %s %s (image=%s, name=%s)", e.Time, e.Type, e.Status, e.ID, e.Image, e.Name)
+	case Image:
+		humanFormat = fmt.Sprintf("%s %s %s %s %s", e.Time, e.Type, e.Status, e.ID, e.Name)
+	}
+	return humanFormat
+}
+
+// NewEventFromString takes stringified json and converts
+// it to an event
+func NewEventFromString(event string) (*Event, error) {
```
Comment (mheon) [inline libpod/events/events.go]: This seems like it ought to be an internal-only method?

File: libpod/events/events.go (hunk 4 above)
Comment (mheon) [reply libpod/events/events.go]: Volumes also do not have IDs

File: libpod/image/image.go [hunk 23]
```diff
@@ -711,7 +717,7 @@ func (i *Image) History(ctx context.Context) ([]*History, error) {
 			Comment:   oci.History[i].Comment,
 		})
 	}
-
+	defer i.newImageEvent(events.History)
```
Comment (mheon) [inline libpod/image/image.go]: Do we need to record this? It's like Inspect - it doesn't modify anything

File: libpod/image/image.go [hunk 24]
```diff
@@ -881,6 +887,7 @@ func (i *Image) Inspect(ctx context.Context) (*inspect.ImageData, error) {
 		User:         ociv1Img.Config.User,
 		History:      ociv1Img.History,
 	}
+	defer i.newImageEvent(events.Inspect)
```
Comment (mheon) [inline libpod/image/image.go]: Again, I really don't think we need an event for this

File: libpod/image/image.go [hunk 25]
```diff
@@ -1148,6 +1159,28 @@ func (i *Image) Save(ctx context.Context, source, format, output string, moreTag
...
+func (ir *Runtime) newImageEvent(status events.Status) {
+	e := events.NewEvent(status)
+	e.Type = events.Image
+	if err := e.Write(ir.EventsLogFilePath); err != nil {
+		logrus.Infof("unable to write event log to %s", ir.EventsLogFilePath)
+	}
+}
+
+// newImageEvent creates a new event based on an image
+func (i *Image) newImageEvent(status events.Status) {
```
Comment (mheon) [inline libpod/image/image.go]: Can't we just make this image.runtime.newImageEvent()

File: libpod/image/prune.go [hunk 26]
```diff
@@ -38,10 +41,16 @@ func (ir *Runtime) PruneImages(all bool) ([]string, error) {
...
 			return nil, errors.Wrap(err, "failed to prune image")
 		}
+		defer p.newImageEvent(events.Prune)
 		prunedCids = append(prunedCids, p.ID())
 	}
 	return prunedCids, nil
 }
+
+// Prune an image
+func (i *Image) Prune() error {
```
Comment (mheon) [inline libpod/image/prune.go]: Why do we need this?

File: libpod/pod_api.go [hunk 27]
```diff
@@ -309,7 +311,7 @@ func (p *Pod) Restart(ctx context.Context) (map[string]error, error) {
 	if len(ctrErrors) > 0 {
 		return ctrErrors, errors.Wrapf(ErrCtrExists, "error stopping some containers")
 	}
-
+	defer p.newPodEvent(events.Restart)
```
Comment (mheon) [inline libpod/pod_api.go]: Would prefer separate Stop and Start events

File: libpod/pod_api.go (hunk 8 above)
Comment (mheon) [reply libpod/pod_api.go]: Agree - defer is definitely slower, and we're always calling at the end of functions anyways

File: libpod/runtime.go [hunk 28]
```diff
@@ -221,6 +221,9 @@ type RuntimeConfig struct {
 	// NumLocks is the number of locks to make available for containers and
 	// pods.
 	NumLocks uint32 `toml:"num_locks,omitempty"`
+
+	//	EventsLogFilePath is where the events log is stored.
```
Comment (mheon) [inline libpod/runtime.go]: Wierd spacing in the comment?

File: libpod/runtime.go [hunk 29]
```diff
@@ -458,7 +461,10 @@ func NewRuntime(options ...RuntimeOption) (runtime *Runtime, err error) {
 			}
 		}
 	}
-
+	runtime.config.EventsLogFilePath = filepath.Join(runtime.config.TmpDir, "events", "events.log")
```
Comment (mheon) [inline libpod/runtime.go]: Why is this in NewRuntime and makeRuntime both? Drop the one in NewRuntime

File: libpod/runtime.go (hunk 9 above)
Comment (mheon) [reply libpod/runtime.go]: I thought we did not want to expose this to anyone - that makes it part of the API, and we need to make guarantees about format? I'd remove this entirely

File: libpod/runtime_ctr.go [hunk 30]
```diff
@@ -239,7 +241,7 @@ func (r *Runtime) newContainer(ctx context.Context, rSpec *spec.Spec, options ..
 func (r *Runtime) RemoveContainer(ctx context.Context, c *Container, force bool, removeVolume bool) error {
 	r.lock.Lock()
 	defer r.lock.Unlock()
-
+	defer c.newContainerEvent(events.Remove)
```
Comment (mheon) [inline libpod/runtime_ctr.go]: Put this in removeContainer

File: libpod/runtime_volume.go [hunk 31]
```diff
@@ -165,13 +167,19 @@ func (r *Runtime) PruneVolumes(ctx context.Context) ([]string, []error) {
...
 			continue
 		}
+		vol.newVolumeEvent(events.Prune)
 		prunedIDs = append(prunedIDs, vol.Name())
 	}
 	return prunedIDs, pruneErrors
 }
+
+// Prune a volume
+func (r *Runtime) Prune(ctx context.Context, vol *Volume, force bool) error {
```
Comment (mheon) [inline libpod/runtime_volume.go]: Why?

File: libpod/runtime_volume.go [hunk 32]
```diff
@@ -34,7 +36,7 @@ func (r *Runtime) RemoveVolume(ctx context.Context, v *Volume, force bool) error
 			return nil
 		}
 	}
-
+	defer v.newVolumeEvent(events.Remove)
```
Comment (mheon) [inline libpod/runtime_volume.go]: Put this in removeVolume

File: libpod/runtime_volume_linux.go [hunk 33]
```diff
@@ -4,6 +4,7 @@ package libpod
 
 import (
 	"context"
+	"github.com/containers/libpod/libpod/events"
```
Comment (mheon) [inline libpod/runtime_volume_linux.go]: Import ordering

File: libpod/runtime_pod_linux.go [hunk 34]
```diff
@@ -118,7 +119,7 @@ func (r *Runtime) NewPod(ctx context.Context, options ...PodCreateOption) (*Pod,
 			return nil, err
 		}
 	}
-
+	pod.newPodEvent(events.Create)
```
Comment (mheon) [inline libpod/runtime_pod_linux.go]: I don't see a remove event for pods

File: libpod/container_api.go (hunk 12 above)
Comment (rhatdan) [reply libpod/container_api.go]: No I don't think we should include this.  Reading info is not an event.  As far as I am concerned.

File: libpod/container_api.go (hunk 15 above)
Comment (rhatdan) [reply libpod/container_api.go]: Just Restore

File: libpod/container_internal.go (hunk 16 above)
Comment (rhatdan) [reply libpod/container_internal.go]: I agree

File: libpod/events.go (hunk 18 above)
Comment (rhatdan) [reply libpod/events.go]: SGTM

File: libpod/image/image.go (hunk 23 above)
Comment (rhatdan) [reply libpod/image/image.go]: Agree no "READING" events

File: libpod/image/image.go (hunk 24 above)
Comment (rhatdan) [reply libpod/image/image.go]: Ditto

File: libpod/pod_api.go (hunk 8 above)
Comment (baude) [reply libpod/pod_api.go]: removingf as you guys dont want inspect anyways

File: libpod/container_internal.go (hunk 16 above)
Comment (baude) [reply libpod/container_internal.go]: docker has restart, seems valid... only we consider restart to be stop/start a user wouldnt

Comment (mheon) [reply libpod/container_internal.go]: Hm. Can we keep restart and have it also log the start/stop that happen?

File: libpod/events/events.go (hunk 22 above)
Comment (baude) [reply libpod/events/events.go]: then libpod/events will not be able to use it.

Comment (mheon) [reply libpod/events/events.go]: Fair enough. Can we say it's internal-only then? Would like to discourage people trying to make their own events with it

File: libpod/image/image.go (hunk 25 above)
Comment (baude) [reply libpod/image/image.go]: there is no image.runtime.  there is image.imageruntime but that leads to image storage not libpod runtime.  image has to be clean of libpod or it will be circular

Comment (mheon) [reply libpod/image/image.go]: But the image runtime has the newImageEvent method too - we can just call that here instead of duplicating the logic

File: libpod/runtime.go (hunk 29 above)
Comment (baude) [reply libpod/runtime.go]: ?? i dont see the one in makeRuntime. mind pointing out?

**Comment by baude** [comment]
> Missing completions.

why do you think that?

```
_podman_events() {
    local options_with_args="
     --help
     --h
     --filter
     --format
     --since
     --until
    "
    case "$cur" in
	-*)
	    COMPREPLY=($(compgen -W "$options_with_args" -- "$cur"))
	    ;;
    esac
}
```

File: libpod/runtime.go (hunk 29 above)
Comment (mheon) [reply libpod/runtime.go]: Sorry, this should have been one line down - we're double-creating the events dir

File: libpod/runtime.go (hunk 9 above)
Comment (baude) [reply libpod/runtime.go]: removed ...

File: libpod/events/events.go (hunk 5 above)
Comment (baude) [reply libpod/events/events.go]: now caught in the function that calls this.

File: docs/podman-events.1.md [hunk 35]
```diff
@@ -0,0 +1,135 @@
...
+Print usage statement.
+
+**--format**
+
+Format the output using the given Go template.  An output value of *json* is not supported.
+
+
+**--filter**=[]
+
+Filter events that are displayed.  The must must be in the format of "filter=value".  The following
```
Comment (TomSweeneyRedHat) [inline docs/podman-events.1.md]: s/must must/must/

File: docs/podman-events.1.md [hunk 36]
```diff
@@ -0,0 +1,135 @@
...
+```
+$ sudo podman events --since 5m
+2019-03-02 10:44:29.548253253 -0600 CST image inspect da86e6ba6ca1 k8s.gcr.io/pause:3.1
+2019-03-02 10:44:29.598835409 -0600 CST container create b629d10d3831 (image=k8s.gcr.io/pause:3.1, name=1df5ebca7b44-infra)
+2019-03-02 10:44:29.601746633 -0600 CST pod create 1df5ebca7b44 (image=, name=confident_hawking)
+2019-03-02 10:44:42.298198458 -0600 CST image inspect da86e6ba6ca1 k8s.gcr.io/pause:3.1
+2019-03-02 10:44:42.371100253 -0600 CST container create 170a0f457d00 (image=k8s.gcr.io/pause:3.1, name=ca731231718e-infra)
+2019-03-02 10:44:42.374637304 -0600 CST pod create ca731231718e (image=, name=webapp)
+```
+
```
Comment (TomSweeneyRedHat) [inline docs/podman-events.1.md]: The examples are great, but it would be nice to have at least one of the three not be a create event.

Comment (TomSweeneyRedHat) [reply docs/podman-events.1.md]: and have one that uses --format.

File: cmd/podman/events.go [hunk 37]
```diff
@@ -0,0 +1,47 @@
...
+		Use:   "events [flags]",
+		Short: "show podman events",
+		Long:  eventsDescription,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			eventsCommand.InputArgs = args
+			eventsCommand.GlobalFlags = MainGlobalOpts
+			return eventsCmd(&eventsCommand)
+		},
+		Example: `podman events
+  podman events -f`,
```
Comment (TomSweeneyRedHat) [inline cmd/podman/events.go]: would be nice to have a little more detailed event here.  I'd keep the first then add the one from your manpage with the filter and what ever you work up for the format.   As a rule of thumb I've been trying to make it 3 examples max.

**Comment by baude** [comment]
bot, retest this please

File: libpod/runtime.go (hunk 29 above)
Comment (baude) [reply libpod/runtime.go]: i have to reverse course on this.  removing this results in the test suite not having eventlog dirs being created.  there must be something about how the tmpdir is being processed.

Comment (mheon) [reply libpod/runtime.go]: Wait, what?

File: cmd/podman/varlink/io.podman.varlink [hunk 38]
```diff
@@ -435,6 +435,24 @@ type Runlabel(
...
+type EventInput(
+    filter: []string,
+    since: string,
+    stream: bool,
+    until: string
+)
+
+# Event describes a libpod struct
+type Event(
+    id: string,
```
Comment (larskarlitski) [inline cmd/podman/varlink/io.podman.varlink]: You can use a `enum` here, which makes it easier for people to see which kinds of events can occur.

File: cmd/podman/varlink/io.podman.varlink [hunk 39]
```diff
@@ -435,6 +435,24 @@ type Runlabel(
     opts: [string]string
 )
 
+# EventInput describes the input to obtain libpod events
+type EventInput(
+    filter: []string,
```
Comment (larskarlitski) [inline cmd/podman/varlink/io.podman.varlink]: If you use an `enum` for `Event.id` (see below), you should use that here as well.

File: cmd/podman/varlink/io.podman.varlink [hunk 40]
```diff
@@ -1120,6 +1138,9 @@ method GetPodsByContext(all: bool, latest: bool, args: []string) -> (pods: []str
 # LoadImage allows you to load an image into local storage from a tarball.
 method LoadImage(name: string, inputFile: string, quiet: bool, deleteFile: bool) -> (reply: MoreResponse)
 
+# GetEvents returns known libpod events filtered by the options provided.
+method GetEvents(options: EventInput) -> (events: Event)
```
Comment (larskarlitski) [inline cmd/podman/varlink/io.podman.varlink]: I assume this only makes sense when it supports `more`? Please note that in the comment.

File: cmd/podman/varlink/io.podman.varlink [hunk 41]
```diff
@@ -1149,3 +1170,9 @@ error ErrorOccurred (reason: string)
 
 # RuntimeErrors generally means a runtime could not be found or gotten.
 error RuntimeError (reason: string)
+
+# The Podman endpoint requires that you use a streaming connection.
+error WantsMoreRequired (reason: string)
+
+# The Podman endpoint has closed because the stream ended.
+error StreamEnded()
```
Comment (larskarlitski) [inline cmd/podman/varlink/io.podman.varlink]: I think the varlink protocol already provides this. Why were these necessary?

File: libpod/runtime.go [hunk 42]
```diff
@@ -221,6 +221,9 @@ type RuntimeConfig struct {
 	// NumLocks is the number of locks to make available for containers and
 	// pods.
 	NumLocks uint32 `toml:"num_locks,omitempty"`
+
+	// EventsLogFilePath is where the events log is stored.
+	EventsLogFilePath string `toml:"events_logfile_path"`
```
Comment (mheon) [inline libpod/runtime.go]: Make this `toml:-` - I don't really want people messing with this via config file

File: cmd/podman/varlink/io.podman.varlink (hunk 40 above)
Comment (baude) [reply cmd/podman/varlink/io.podman.varlink]: indeed, fixed.

File: cmd/podman/events.go [hunk 43]
```diff
@@ -0,0 +1,48 @@
...
+		Short: "show podman events",
+		Long:  eventsDescription,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			eventsCommand.InputArgs = args
+			eventsCommand.GlobalFlags = MainGlobalOpts
+			return eventsCmd(&eventsCommand)
+		},
+		Example: `podman events
+podman events --filter event=create
+podman events --since 1h30s`,
```
Comment (TomSweeneyRedHat) [inline cmd/podman/events.go]: add two spaces in front of this line and the one above please.

File: cmd/podman/varlink/io.podman.varlink (hunk 38 above)
Comment (larskarlitski) [reply cmd/podman/varlink/io.podman.varlink]: To clarify, I suggest making an `enum` type like this:
```
type EventType (container, image, foo, baz)
```
And then use it here and for the `filter` property. Everything on the wire stays the same.

File: cmd/podman/varlink/io.podman.varlink (hunk 40 above)
Comment (larskarlitski) [reply cmd/podman/varlink/io.podman.varlink]: Also, please put the fields of the `EventInput` type into this call. It's awkward to call otherwise.

File: cmd/podman/varlink/io.podman.varlink (hunk 38 above)
Comment (mheon) [reply cmd/podman/varlink/io.podman.varlink]: I don't see how `ID` and `filter` make any sense as enums - ID is the container/pod ID, and makes no sense. Filters are arguable, but as we're sending them now a string seems to make more sense (so we can encode user-requested values to filter against)

**Comment by martinpitt** [comment]
I'm interested in understanding the behaviour here.

You mention Cockpit in the description, but the description sounds like it would add logging and display past events, not future ones (the bit that UIs are interested in). Or is `podman events` a blocking command that shows you new events as long as it's running? (i. e. a varlink call with `more`, which behaves like a signal notification)?

Thanks!

**Comment by rhatdan** [comment]
The goal is to have that correct.  Think of it like tail -f of the events log.

**Comment by martinpitt** [comment]
@rhatdan : Perfect, thank you!

**Comment by rhatdan** [comment]
@martinpitt BTW You are the driving factor in adding this feature, so we want to make sure that it satisfies your needs/requirements.

File: libpod/events.go [hunk 44]
```diff
@@ -0,0 +1,81 @@
...
+import (
+	"github.com/containers/libpod/libpod/events"
+	"github.com/hpcloud/tail"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// newContainerEvent creates a new event based on a container
+func (c *Container) newContainerEvent(status events.Status) {
+	e := events.NewEvent(status)
```
Comment (rhatdan) [inline libpod/events.go]: Shouldn't events.NewEvent just take these fields as  paramaters?

File: libpod/events.go [hunk 45]
```diff
@@ -0,0 +1,81 @@
...
+	e.ID = c.ID()
+	e.Name = c.Name()
+	e.Image = c.config.RootfsImageName
+	e.Type = events.Container
+	if err := e.Write(c.runtime.config.EventsLogFilePath); err != nil {
+		logrus.Errorf("unable to write event log to %s", c.runtime.config.EventsLogFilePath)
```
Comment (rhatdan) [inline libpod/events.go]: This is not a log, I would just say unable to write event to ...

File: libpod/events.go [hunk 46]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+// newPodEvent creates a new event for a libpod pod
+func (p *Pod) newPodEvent(status events.Status) {
+	e := events.NewEvent(status)
+	e.ID = p.ID()
+	e.Name = p.Name()
+	e.Type = events.Pod
+	if err := e.Write(p.runtime.config.EventsLogFilePath); err != nil {
+		logrus.Errorf("unable to write event log to %s", p.runtime.config.EventsLogFilePath)
```
Comment (rhatdan) [inline libpod/events.go]: remove log

File: libpod/events.go [hunk 47]
```diff
@@ -0,0 +1,81 @@
...
+	}
+}
+
+// newVolumeEvent creates a new event for a libpod volume
+func (v *Volume) newVolumeEvent(status events.Status) {
+	e := events.NewEvent(status)
+	e.Name = v.Name()
+	e.Type = events.Volume
+	if err := e.Write(v.runtime.config.EventsLogFilePath); err != nil {
+		logrus.Errorf("unable to write event log to %s", v.runtime.config.EventsLogFilePath)
```
Comment (rhatdan) [inline libpod/events.go]: Remove log.

File: cmd/podman/shared/events.go [hunk 48]
```diff
@@ -0,0 +1,115 @@
...
+				return true
+			}
+			return strings.HasPrefix(e.ID, filterValue)
+		}, nil
+	case "VOLUME":
+		return func(e *events.Event) bool {
+			if e.Type != events.Volume {
+				return false
+			}
+			return strings.HasPrefix(e.ID, filterValue)
```
Comment (TomSweeneyRedHat) [inline cmd/podman/shared/events.go]: just checking, do you need an `if e.name == filtervalue {` statement here for Volumes?

File: docs/podman-events.1.md [hunk 49]
```diff
@@ -0,0 +1,139 @@
...
+$ podman events --filter event=create
+2019-03-02 10:36:01.375685062 -0600 CST container create 20dc581f6fbf (image=docker.io/library/alpine:latest, name=sharp_morse)
+2019-03-02 10:36:08.561188337 -0600 CST container create 58e7e002344c (image=k8s.gcr.io/pause:3.1, name=3e701f270d54-infra)
+2019-03-02 10:36:13.146899437 -0600 CST volume create cad6dc50e087 (image=, name=cad6dc50e0879568e7d656bd004bd343d6035e7fc4024e1711506fe2fd459e6f)
+2019-03-02 10:36:29.978806894 -0600 CST container create d81e30f1310f (image=docker.io/library/busybox:latest, name=musing_newton)
+```
+
+Show only podman pod create events
+```
+$ podman events --filter event=create --filter type=pod
```
Comment (TomSweeneyRedHat) [inline docs/podman-events.1.md]: It would be nice to have an event other than create for this example.  Ditto an example showing the --format command in play.

File: cmd/podman/varlink/io.podman.varlink (hunk 38 above)
Comment (larskarlitski) [reply cmd/podman/varlink/io.podman.varlink]: Sorry I confused the fields (thought `id` is an event id). I meant status.

File: cmd/podman/shared/events.go (hunk 48 above)
Comment (baude) [reply cmd/podman/shared/events.go]: volumes dont have names

File: libpod/events.go (hunk 44 above)
Comment (baude) [reply libpod/events.go]: this is way more convenient for adding the events later. and most of it can be derived from the method object (container, pod...)

File: cmd/podman/shared/events.go (hunk 48 above)
Comment (mheon) [reply cmd/podman/shared/events.go]: No, volumes only have names, not IDs

**Comment by baude** [comment]
ready for merge
